File: mcp-list/packages/EventAnalyzer/src/event_analyzer.py
"""Event Analyzer
分析埋点数据，检测字段问题
"""

from typing import Dict, Any, List
from src.utils.type_checker import TypeChecker, TypeName


class EventAnalyzer:
    """埋点事件分析器"""

    def __init__(self):
        self.type_checker = TypeChecker()

    def analyze(
        self,
        event_data: Dict[str, Any],
        field_definitions: Dict[str, Dict],
        check_required: bool = False
    ) -> Dict[str, Any]:
        """
        分析埋点数据，检测问题

        Args:
            event_data: 埋点数据
            field_definitions: 字段定义（从 API 获取）
            check_required: 是否检查必填字段

        Returns:
            分析结果
        """
        issues = []
        properties = event_data.get("properties", {})

        # 1. 检查类型匹配
        for field_name, value in properties.items():
            if field_name in field_definitions:
                expected_type = field_definitions[field_name]["type"]
                actual_type = self.type_checker.infer_type(value)

                if not self.type_checker.type_matches(expected_type, actual_type):
                    issues.append({
                        "type": "type_mismatch",
                        "field": field_name,
                        "expected": expected_type,
                        "actual": actual_type,
                        "value": value,
                        "severity": "error",
                        "message": f"字段 {field_name} 类型错误: 期望 {expected_type}, 实际 {actual_type}"
                    })

        # 2. 检查字段是否在定义中（可能是拼写错误或废弃字段）
        for field_name in properties.keys():
            if field_name not in field_definitions:
                # 跳过系统字段（以 $ 开头）
                if not field_name.startswith("$"):
                    issues.append({
                        "type": "unknown_field",
                        "field": field_name,
                        "severity": "warning",
                        "message": f"字段 {field_name} 不在字段定义中，可能是拼写错误或废弃字段"
                    })

        # 3. 检查枚举值（如果有 trans 定义）
        for field_name, value in properties.items():
            if field_name in field_definitions:
                field_def = field_definitions[field_name]
                if field_def.get("trans"):
                    try:
                        import json
                        enum_values = json.loads(field_def["trans"])
                        str_value = str(value)
                        if str_value not in enum_values:
                            issues.append({
                                "type": "invalid_enum",
                                "field": field_name,
                                "value": value,
                                "valid_values": list(enum_values.keys()),
                                "severity": "warning",
                                "message": f"字段 {field_name} 的值 {value} 不在枚举值范围内"
                            })
                    except Exception:
                        pass

        # 统计信息
        total_fields = len(field_definitions)
        present_fields = len([f for f in properties.keys() if f in field_definitions])

        # 确定状态
        error_count = len([i for i in issues if i.get("severity") == "error"])
        warning_count = len([i for i in issues if i.get("severity") == "warning"])

        if error_count > 0:
            status = "error"
        elif warning_count > 0:
            status = "warning"
        else:
            status = "ok"

        return {
            "event": event_data.get("event", "Unknown"),
            "status": status,
            "summary": f"检测到 {error_count} 个错误, {warning_count} 个警告",
            "issues": issues,
            "fields_present": present_fields,
            "fields_total": total_fields,
            "coverage": f"{present_fields}/{total_fields} ({present_fields*100//total_fields if total_fields > 0 else 0}%)"
        }

    def get_missing_fields(
        self,
        event_data: Dict[str, Any],
        field_definitions: Dict[str, Dict]
    ) -> List[str]:
        """
        获取缺失的字段列表

        Args:
            event_data: 埋点数据
            field_definitions: 字段定义

        Returns:
            缺失的字段名称列表
        """
        properties = event_data.get("properties", {})
        all_fields = set(field_definitions.keys())
        present_fields = set(properties.keys())

        # 排除系统字段（以 $ 开头）
        present_fields = {f for f in present_fields if not f.startswith("$")}

        return sorted(list(all_fields - present_fields))

    def compare_events(
        self,
        event1_fields: Dict[str, Dict],
        event2_fields: Dict[str, Dict]
    ) -> Dict[str, Any]:
        """
        比较两个事件的字段差异

        Args:
            event1_fields: 事件1的字段定义
            event2_fields: 事件2的字段定义

        Returns:
            差异分析结果
        """
        fields1 = set(event1_fields.keys())
        fields2 = set(event2_fields.keys())

        common = fields1 & fields2
        only_in_event1 = fields1 - fields2
        only_in_event2 = fields2 - fields1

        return {
            "common_fields": sorted(list(common)),
            "event1_only": sorted(list(only_in_event1)),
            "event2_only": sorted(list(only_in_event2)),
            "total_common": len(common),
            "total_diff": len(only_in_event1) + len(only_in_event2)
        }

File: mcp-list/packages/EventAnalyzer-node/src/services/eventAnalyzer.ts
/**
 * 事件分析器
 * 分析埋点数据，检测类型错误、未知字段、枚举值错误等
 */

import { TypeChecker, TypeName } from '../utils/typeChecker.js';
import { FieldDefinitions, FieldDefinition } from './apiClient.js';

export interface TypeErrorItem {
  field: string;
  expected: string;
  actual: string;
  value: any;
}

export interface EnumErrorItem {
  field: string;
  value: any;
  valid_values: string[];
}

export interface UnknownFieldItem {
  field: string;
  value: any;
}

export interface MissingFieldItem {
  field: string;
  type: string;
}

export interface AnalysisResult {
  event: string;
  total_fields: number;
  analyzed_fields: number;
  type_errors: TypeErrorItem[];
  enum_errors: EnumErrorItem[];
  unknown_fields: UnknownFieldItem[];
  missing_required_fields?: MissingFieldItem[];
  coverage: string;
  summary: string;
}

export interface ComparisonResult {
  event1: string;
  event2: string;
  common_fields: string[];
  only_in_event1: string[];
  only_in_event2: string[];
  type_differences: Array<{
    field: string;
    event1_type: string;
    event2_type: string;
  }>;
}

export class EventAnalyzer {
  /**
   * 分析埋点数据
   * @param eventData 埋点数据
   * @param fieldDefinitions 字段定义
   * @param checkRequired 是否检查必填字段
   */
  analyze(
    eventData: any,
    fieldDefinitions: FieldDefinitions,
    checkRequired: boolean = false
  ): AnalysisResult {
    const eventName = eventData.event || 'Unknown';
    const typeErrors: TypeErrorItem[] = [];
    const enumErrors: EnumErrorItem[] = [];
    const unknownFields: UnknownFieldItem[] = [];
    const missingRequiredFields: MissingFieldItem[] = [];

    // 1. 检查未知字段
    for (const fieldName in eventData) {
      if (!fieldDefinitions[fieldName]) {
        unknownFields.push({
          field: fieldName,
          value: eventData[fieldName],
        });
      }
    }

    // 2. 检查已定义字段
    for (const fieldName in fieldDefinitions) {
      const fieldDef = fieldDefinitions[fieldName];
      const fieldValue = eventData[fieldName];

      // 检查必填字段
      if (checkRequired && fieldDef.required && (fieldValue === undefined || fieldValue === null)) {
        missingRequiredFields.push({
          field: fieldName,
          type: fieldDef.type,
        });
        continue;
      }

      // 如果字段存在，检查类型和枚举值
      if (fieldValue !== undefined && fieldValue !== null) {
        // 类型检查
        const expectedType = this.mapTypeToTypeName(fieldDef.type);
        const actualType = TypeChecker.inferType(fieldValue);

        if (!TypeChecker.typeMatches(expectedType, actualType, fieldValue)) {
          typeErrors.push({
            field: fieldName,
            expected: TypeChecker.getTypeName(expectedType),
            actual: TypeChecker.getTypeName(actualType),
            value: fieldValue,
          });
        }

        // 枚举值检查
        if (fieldDef.trans) {
          const enumValues = this.parseEnumValues(fieldDef.trans);
          if (!TypeChecker.validateEnum(fieldValue, enumValues)) {
            enumErrors.push({
              field: fieldName,
              value: fieldValue,
              valid_values: Object.keys(enumValues),
            });
          }
        }
      }
    }

    // 3. 计算覆盖率
    const totalFields = Object.keys(fieldDefinitions).length;
    const analyzedFields = totalFields - missingRequiredFields.length;
    const coverage = totalFields > 0 ? ((analyzedFields / totalFields) * 100).toFixed(2) : '0.00';

    // 4. 生成摘要
    const summary = this.generateSummary(
      typeErrors.length,
      enumErrors.length,
      unknownFields.length,
      missingRequiredFields.length
    );

    const result: AnalysisResult = {
      event: eventName,
      total_fields: totalFields,
      analyzed_fields: analyzedFields,
      type_errors: typeErrors,
      enum_errors: enumErrors,
      unknown_fields: unknownFields,
      coverage: `${coverage}%`,
      summary,
    };

    if (checkRequired) {
      result.missing_required_fields = missingRequiredFields;
    }

    return result;
  }

  /**
   * 比较两个事件的字段差异
   */
  compareEvents(fields1: FieldDefinitions, fields2: FieldDefinitions): ComparisonResult {
    const fields1Keys = new Set(Object.keys(fields1));
    const fields2Keys = new Set(Object.keys(fields2));

    const commonFields = [...fields1Keys].filter((key) => fields2Keys.has(key));
    const onlyInEvent1 = [...fields1Keys].filter((key) => !fields2Keys.has(key));
    const onlyInEvent2 = [...fields2Keys].filter((key) => !fields1Keys.has(key));

    // 检查类型差异
    const typeDifferences = commonFields
      .map((field) => {
        const type1 = fields1[field].type;
        const type2 = fields2[field].type;
        if (type1 !== type2) {
          return {
            field,
            event1_type: type1,
            event2_type: type2,
          };
        }
        return null;
      })
      .filter((item) => item !== null) as Array<{
      field: string;
      event1_type: string;
      event2_type: string;
    }>;

    return {
      event1: '',
      event2: '',
      common_fields: commonFields,
      only_in_event1: onlyInEvent1,
      only_in_event2: onlyInEvent2,
      type_differences: typeDifferences,
    };
  }

  /**
   * 映射 API 类型到 TypeName
   */
  private mapTypeToTypeName(apiType: string): TypeName {
    const typeMap: Record<string, TypeName> = {
      NUMBER: 'NUMBER',
      STRING: 'STRING',
      BOOL: 'BOOL',
      LIST: 'LIST',
      OBJECT: 'OBJECT',
      INT: 'NUMBER',
      FLOAT: 'NUMBER',
      BOOLEAN: 'BOOL',
      ARRAY: 'LIST',
      JSON: 'OBJECT',
    };

    const upperType = apiType.toUpperCase();
    return typeMap[upperType] || 'UNKNOWN';
  }

  /**
   * 解析枚举值字符串
   */
  private parseEnumValues(transStr: string): Record<string, string> {
    const result: Record<string, string> = {};

    try {
      const items = transStr.split(',');
      for (const item of items) {
        const [key, value] = item.split(':');
        if (key && value) {
          result[key.trim()] = value.trim();
        }
      }
    } catch {
      // 解析失败
    }

    return result;
  }

  /**
   * 生成分析摘要
   */
  private generateSummary(
    typeErrorCount: number,
    enumErrorCount: number,
    unknownFieldCount: number,
    missingFieldCount: number
  ): string {
    const issues: string[] = [];

    if (typeErrorCount > 0) {
      issues.push(`${typeErrorCount} 个类型错误`);
    }
    if (enumErrorCount > 0) {
      issues.push(`${enumErrorCount} 个枚举值错误`);
    }
    if (unknownFieldCount > 0) {
      issues.push(`${unknownFieldCount} 个未知字段`);
    }
    if (missingFieldCount > 0) {
      issues.push(`${missingFieldCount} 个缺失必填字段`);
    }

    if (issues.length === 0) {
      return '✅ 数据完全符合规范';
    }

    return `⚠️ 发现问题：${issues.join('，')}`;
  }
}
